### ISSUE TITLE
Compile times on Visual Studio 2012 slow

### ISSUE THREAD

**Issue by elliotwoods** [openframeworks/openFrameworks#2548]
Currently our compile times on Windows are pretty slow.

I'm wondering if either @leocolomb might have some thoughts on this, or if @obviousjim could bend an ear at Microsoft to give some suggestions.

Each individual file is slow to compile, and VS doesn't compile in parallel as much as XCode (it compiles seperate projects in parallel within a solution, but not the files within the project). And I presume the duration is simply because there are so many headers associated with each file (so essentially each source file is enormous by the time it makes it the compiler).

I presume the official line is something like 'use precompiled headers!', and i'm curious what is involved in implementing that.

Maybe LLVM/GCC is doing some magic to automatically speed up the process (e.g. auto-precompiling headers), and perhaps we need to just squeeze MSVC in the right way to get it to act in the same way. Or VS2013 might bring along magic fixes.

I haven't had any positive experiences with precompiled headers yet, and am not pressing for them to be a core strategy.

**Comment by LeoColomb** [comment]
Fixed by #3572 ? (and move to #255)

**Comment by bilderbuchi** [comment]
I don't know (not a Win user), what's your opinion?

**Comment by LeoColomb** [comment]
I don't know what is the targeted time...
With multiprocessor tag, the compile time is incredibly improved, but still around 1 minute.

Nevertheless, @elliotwoods's original remark is now fixed.

> Each individual file is slow to compile, and VS doesn't compile in parallel as much as XCode

**Comment by admsyn** [comment]
IMO this can be closed, unless there's any other low-hanging fruit like the compiler flag? Especially given that vs2012 is presumably on the way out.

**Comment by bilderbuchi** [comment]
alright then :-)

**Comment by elliotwoods** [comment]
revisiting this. i'm looking into precompiled headers again
multi-processor compilation definitely helps! but i think good use of precompiled headers still might save the 8 tonnes of carbon dioxide being emitted each year from people compiling cpp's in oF/VS 
will make a new issue once i understand what's what / do some tests

**Comment by elliotwoods** [comment]
Got this working in Visual Studio also

For a 2 file (main.cpp, ofApp.cpp) project with 1 line of code in draw
VS2015 / Xeon 6-core / x64 / Debg build:

Using PCH:

`
1>      283 ms  ClCompile                                  1 calls
1>     3408 ms  Link                                       1 calls
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
`

Without 

`
1>     3535 ms  ClCompile                                  1 calls
1>     4977 ms  Link                                       1 calls
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
`

That's a pretty drastic increase in speed for the compile stage. Linking is still quite heavy, but for larger projects (more than 2 cpp's) the effect will be more pronounced.

**Comment by danoli3** [comment]
What does the PCH contain references to?
This could help all environments with Core Lib build times.

+1 to this idea

**Comment by kritzikratzi** [comment]
i haven't had the time to try this yet, but setting up [clcache](https://github.com/frerich/clcache) locally could be an alternative to pch (i've read in some places that pch is discouraged now in xcode).
